lib/pos_mempool.go

```diff
@@ -0,0 +1,391 @@
+
+	return dmp.removeTransactionNoLock(txn)
+}
+
+func (dmp *PosMempool) removeTransactionNoLock(txn *MempoolTx) error {
+	// First, sanity check our reserved balance.
+	userPk := NewPublicKey(txn.Tx.PublicKey)
+
+	// Remove the transaction from the register.
+	if err := dmp.txnRegister.RemoveTransaction(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.RemoveTransaction: Problem removing txn from register")
+	}
+	// Decrease the appropriate ledger's balance by the transaction fee.
+	dmp.ledger.DecreaseBalance(*userPk, txn.Fee)
+
+	// Emit an event for the removed transaction.
+	if dmp.emitEvents {
+		event := &MempoolEvent{
+			Txn:  txn,
+			Type: MempoolEventRemove,
+		}
+		dmp.persister.EnqueueEvent(event)
+	}
+
+	return nil
+}
+
+// GetTransaction returns the transaction with the given hash if it exists in the mempool. This function is thread-safe.
+func (dmp *PosMempool) GetTransaction(txHash *BlockHash) *MempoolTx {
+	if !dmp.IsRunning() {
+		return nil
+	}
+
+	dmp.RLock()
+	defer dmp.RUnlock()
+
+	return dmp.txnRegister.GetTransaction(txHash)
+}
+
+// GetTransactions returns all transactions in the mempool ordered by the Fee-Time algorithm. This function is thread-safe.
+func (dmp *PosMempool) GetTransactions() []*MempoolTx {
+	if !dmp.IsRunning() {
+		return nil
+	}
+
+	dmp.RLock()
+	defer dmp.RUnlock()
+
+	return dmp.txnRegister.GetFeeTimeTransactions()
+}
+
+// GetIterator returns an iterator for the mempool transactions. The iterator can be used to peek transactions in the
+// mempool ordered by the Fee-Time algorithm. Transactions can be fetched with the following pattern:
+//
+//	for it.Next() {
+//		if txn, ok := it.Value(); ok {
+//			// Do something with txn.
+//		}
+//	}
+//
+// Note that the iteration pattern is not thread-safe. Another lock should be used to ensure thread-safety.
+func (dmp *PosMempool) GetIterator() MempoolIterator {
+	if !dmp.IsRunning() {
+		return nil
+	}
+
+	dmp.RLock()
+	defer dmp.RUnlock()
+
+	return dmp.txnRegister.GetFeeTimeIterator()
+}
+
+// pruneNoLock removes transactions from the mempool until the mempool size is below the maximum allowed size. The transactions
+// are removed in lowest to highest Fee-Time priority, i.e. opposite way that transactions are ordered in
+// GetTransactions().
+func (dmp *PosMempool) pruneNoLock() error {
+	if dmp.txnRegister.Size() < dmp.params.MaxMempoolPosSizeBytes {
+		return nil
+	}
+
+	prunedTxns, err := dmp.txnRegister.PruneToSize(dmp.params.MaxMempoolPosSizeBytes)
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.prune: Problem pruning mempool")
+	}
+	for _, prunedTxn := range prunedTxns {
+		if err := dmp.removeTransactionNoLock(prunedTxn); err != nil {
+			glog.Errorf("PosMempool.prune: Problem removing transaction from mempool: %v", err)
+		}
+	}
+	return nil
+}
+
+// UpdateLatestBlock updates the latest block view and latest block node in the mempool.
+func (dmp *PosMempool) UpdateLatestBlock(blockView *UtxoView, blockNode *BlockNode) {
```

Comment:
Same comment as above. May as well just pass in the block height since that's the piece of info we care about
```suggestion
func (dmp *PosMempool) UpdateLatestBlock(blockView *UtxoView, blockHeight uint64) {
```